File: 445-add-two-numbers-ii/445-add-two-numbers-ii.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode addTwoNumbers(ListNode l1, ListNode l2) {
        Stack<Integer> s1 = new Stack<>();
        Stack<Integer> s2 = new Stack<>();
        while(l1 != null) {
            s1.push(l1.val);
            l1 = l1.next;
        }
        while(l2 != null) {
            s2.push(l2.val);
            l2 = l2.next;
        }
        int carry = 0;
        ListNode ans = null;
        while(!s1.isEmpty() || !s2.isEmpty() || carry != 0 ) {
            int x = s1.isEmpty() ? 0 : s1.pop();
            int y = s2.isEmpty() ? 0 : s2.pop();
            ListNode node = new ListNode((x + y + carry) % 10);
            node.next = ans;
            ans = node;
            carry = (x + y + carry) / 10;
        }
        return ans;
    }
}

File: 445-add-two-numbers-ii/445-add-two-numbers-ii.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def addTwoNumbers(self, l1: ListNode, l2: ListNode) -> ListNode:
        s1, s2= [], []
        
        while l1:
            s1.append(l1.val)
            l1 = l1.next
        while l2:
            s2.append(l2.val)
            l2 = l2.next
            
        carry = 0
        dummy = ListNode(-1)
        dummy.next = None
        
        while s1 or s2:
            val1 = s1.pop() if s1 else 0
            val2 = s2.pop() if s2 else 0
            sum = val1 + val2 +carry
            node = ListNode(sum % 10)
            carry = sum // 10
            node.next = dummy.next
            dummy.next = node
        if carry:
            node= ListNode(carry)
            node.next = dummy.next
            dummy.next = node
        return dummy.next